lib/chef/provisioning/aws_driver/aws_provider.rb

```diff
@@ -22,8 +23,184 @@ def region
+      if aws_object
+        Chef::Log.debug "#{new_resource.to_s} provided #{new_resource.class.aws_id_attribute} #{desired_id} in #{desired_driver.driver_url}.  Will delete."
+      end
+    end
+
+    #
+    # Managed entries are looked up by ID.
+    #
+    if new_resource.is_a?(AWSResourceWithEntry)
+      entry_driver, entry_id, entry = new_resource.get_id_from_managed_entry
+      if entry_id
+        if desired_id && (desired_id != entry_id || desired_driver.driver_url != entry_driver.driver_url)
+          if new_resource.class.get_aws_object(entry_id, driver: entry_driver, resource: new_resource, required: false)
+            # If the desired ID / driver differs from the entry, don't delete.  We
+            # certainly can't delete the AWS object itself, and we don't *want* to
+            # delete the association, because the expectation is that after doing a
+            # delete, you should be able to create a new thing.
+            raise "#{new_resource.to_s} provided #{new_resource.class.aws_id_attribute} #{desired_id} in #{desired_driver.driver_url}, but is currently associated with #{entry_id} in #{entry_driver.driver_url}.  Cannot delete the entry or the association until this inconsistency is resolved."
+          else
+            Chef::Log.debug "#{new_resource.to_s} provided #{new_resource.class.aws_id_attribute} #{desired_id} in #{desired_driver.driver_url}, but is currently associated with #{entry_id} in #{entry_driver.driver_url}, which does not exist.  Will delete #{desired_id} and disassociate from #{entry_id}."
+          end
+        else
+
+          # Normal case: entry exists, and is the same as desired (or no desired)
+          aws_object = new_resource.class.get_aws_object(entry_id, driver: entry_driver, resource: new_resource, required: false)
+          if aws_object
+            Chef::Log.debug "#{new_resource.to_s} is associated with #{entry_id} in #{entry_driver.driver_url}.  Will delete."
+          else
+            Chef::Log.debug "#{new_resource.to_s} is associated with #{entry_id} in #{entry_driver.driver_url}, but it does not exist.  Will disassociate the entry but not delete."
+          end
+        end
+      end
+
+    #
+    # Non-managed entries all have their own way of looking it up
+    #
+    else
+      aws_object ||= new_resource.aws_object
+    end
+
+    #
+    # Call the delete method
+    #
+    if aws_object
+      destroy_aws_object(aws_object)
+    end
+
+    #
+    # Associate the managed entry with the AWS object
+    #
+    if new_resource.is_a?(AWSResourceWithEntry) && entry
+      new_resource.delete_managed_entry(action_handler)
+    end
+  end
+
+  protected
+
+  def create_aws_object
+    raise NotImplementedError, :create_aws_object
+  end
+
+  def update_aws_object(obj)
+    raise NotImplementedError, :update_aws_object
+  end
+
+  def destroy_aws_object(obj)
+    raise NotImplementedError, :destroy_aws_object
```

Comment:
It seems like this could actually have a default implementation of 
```ruby
converge_by "Destroying #{obj.class} #{obj.name} (#{obj.aws_id})" do
  obj.destroy
end
```